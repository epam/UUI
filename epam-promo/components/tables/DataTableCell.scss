@use '../../assets/styles/colors' as *;

@mixin checkbox-position($value) {
    .after-paddings, .before-paddings {
        :global(.uui-checkbox), .folding-arrow,
        :global(.uui-checkbox) + :global(.uui-icon) {
            margin-top: $value;
        }
    }
}

.cell {
    position: relative;
    box-sizing: border-box;

    &.wrapper {
        display: flex;
        align-items: flex-start;
    }

    &.align-widgets-top {
        &.size-24, &.size-30 {
           @include checkbox-position(6px);
        }

        &.size-36 {
            @include checkbox-position(9px);
        }

        &.size-42 {
            @include checkbox-position(12px);
        }

        &.size-48 {
            @include checkbox-position(15px);
        }

        &.size-60 {
            @include checkbox-position(21px);
        }
    }

    &.align-widgets-center {
        .after-paddings, .before-paddings {
            align-self: center;
        }
    }

    &.padding-12 {
        padding: 0 12px;
    }

    &.padding-24 {
        padding: 0 24px;
    }

    &.padding-left-12 {
        padding-left: 12px;
    }

    &.padding-left-18 {
        padding-left: 18px;
    }

    &.padding-left-24 {
        padding-left: 24px;
    }

    &.padding-left-30 {
        padding-left: 30px;
    }

    &.padding-left-36 {
        padding-left: 36px;
    }

    &.padding-right-24 {
        padding-right: 24px;
    }
}

.wrapped-content {
    flex: 1 1 0;
    min-width: 0;
    padding-left: 12px;
}

.drag-handle {
    position: absolute;
    left: 2px;
    width: 6px;
    top: 3px;
    bottom: 2px;
    align-self: stretch;
    visibility: hidden;

    :global(.-draggable.uui-table-row):not(:global(.uui-drop-accepted)):hover & {
        visibility: visible;
    }
}

.indent {
    display: flex;
    justify-content: flex-start;
    flex-grow: 0;
    flex-shrink: 0;
    width: 12px;
    padding-left: 12px;
}

.folding-arrow {
    margin: auto 0;

    &:global(.uui-icon) {
        position: relative;
        fill: $gray60;
    }
}

.folding-arrow-12 {
    &:global(.uui-icon) {
        left: -2px;
        flex: 0 0 12px;
        width: 12px;
        height: 12px;
    }
}

.folding-arrow-18 {
    &:global(.uui-icon) {
        left: -4px;
        flex: 0 0 18px;
        width: 18px;
        height: 18px;
    }
}

.before-paddings {
    box-sizing: border-box;
    display: flex;
    justify-content: space-around;
    position: absolute;
    height: 100%;
    left: 0;
    top: 0;

    .indent {
        justify-content: center;
    }

    .folding-arrow {
        &:global(.uui-icon) {
            left: 0;
        }
    }
}

.after-paddings {
    display: flex;
    justify-content: space-around;
}
