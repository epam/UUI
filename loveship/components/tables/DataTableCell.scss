@use '../../assets/styles/scss/index' as *;

@mixin checkbox-position($value) {
    .after-paddings, .before-paddings {
        :global(.uui-checkbox), .folding-arrow,
        :global(.uui-checkbox) + :global(.uui-icon) {
            margin-top: $value;
        }
    }
}

.cell {
    position: relative;
    box-sizing: border-box;

    &.wrapper {
        display: flex;
        align-items: flex-start;
    }

    &.align-widgets-top {
        &.size-24, &.size-30 {
           @include checkbox-position(6px);
        }

        &.size-36 {
            @include checkbox-position(9px);
        }

        &.size-42 {
            @include checkbox-position(12px);
        }

        &.size-48 {
            @include checkbox-position(15px);
        }

        &.size-60 {
            @include checkbox-position(21px);
        }
    }

    &.align-widgets-center {
        .after-paddings, .before-paddings {
            align-self: center;
        }
    }

    &.padding-12 { padding: 0 12px; }
    &.padding-24 { padding: 0 24px; }

    &.padding-left-12 { padding-left: 12px; }
    &.padding-left-18 { padding-left: 18px; }
    &.padding-left-24 { padding-left: 24px; }
    &.padding-left-30 { padding-left: 30px; }
    &.padding-left-36 { padding-left: 36px; }

    &.padding-right-12 { padding-right: 12px; }
    &.padding-right-18 { padding-right: 18px; }
    &.padding-right-24 { padding-right: 24px; }
    &.padding-right-30 { padding-right: 30px; }
    &.padding-right-36 { padding-right: 36px; }

    &.vPadding-6 {
        .before-paddings, & {
            padding-top: 3px;
            padding-bottom: 3px;
        }

    }

    &.vPadding-12 {
        .before-paddings, & {
            padding-top: 6px;
            padding-bottom: 6px;
        }
    }

    &.vPadding-18 {
        .before-paddings, & {
            padding-top: 9px;
            padding-bottom: 9px;
        }
    }

    &.vPadding-24 {
        .before-paddings, & {
            padding-top: 12px;
            padding-bottom: 12px;
        }
    }

    &.vPadding-30 {
        .before-paddings, & {
            padding-top: 15px;
            padding-bottom: 15px;
        }
    }

    &.vPadding-36 {
        .before-paddings, & {
            padding-top: 18px;
            padding-bottom: 18px;
        }
    }
}

.wrapped-content {
    flex: 1 1 0;
    min-width: 0;
}

.drag-handle {
    position: absolute;
    left: 2px;
    width: 6px;
    top: 3px;
    bottom: 2px;
    align-self: stretch;
    visibility: hidden;

    :global(.-draggable.uui-table-row):not(:global(.uui-drop-accepted)):hover & {
        visibility: visible;
    }
}

.indent {
    display: flex;
    justify-content: flex-start;
    flex-grow: 0;
    flex-shrink: 0;
    width: 24px;
}

.label {
    position: absolute;
    width: 6px;
    left: 0;
    top: 0;
    height: 100%;
    background: var(--uui-color-700);
}

.folding-arrow {
    margin: auto 0;

    &:global(.uui-icon) {
        position: relative;
        fill: $night600;
    }
}

.folding-arrow-12 {
    &:global(.uui-icon) {
        left: -2px;
        flex: 0 0 12px;
        width: 12px;
        height: 12px;
    }
}

.folding-arrow-18 {
    &:global(.uui-icon) {
        left: -4px;
        flex: 0 0 18px;
        width: 18px;
        height: 18px;
    }
}

.before-paddings {
    box-sizing: border-box;
    display: flex;
    justify-content: space-around;
    position: absolute;
    height: 100%;
    left: 0;
    top: 0;

    .indent {
        justify-content: center;
    }

    .folding-arrow {
        &:global(.uui-icon) {
            left: 0;
        }
    }
}

.after-paddings {
    display: flex;
    justify-content: space-around;

    .checkbox {
        margin-right: 12px;
    }
}
