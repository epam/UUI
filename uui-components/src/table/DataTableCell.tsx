import React from 'react';
import { DataTableCellProps, RenderEditorProps, uuiElement, uuiMod } from '@epam/uui-core';
import css from './DataTableCell.scss';
import { FlexCell } from '../layout';
import { DataTableCellOverlay } from './DataTableCellOverlay';

interface DataTableCellState {
    inFocus: boolean;
}

const uuiDataTableCellMarkers = {
    uuiTableCell: 'uui-table-cell',
} as const;

export const DataTableCell = <TItem, TId, TCellValue>(props: DataTableCellProps<TItem, TId, TCellValue>) => {
    const [state, setState] = React.useState<DataTableCellState>({ inFocus: false });
    const row = props.rowProps;
    const ref = React.useRef<HTMLDivElement>();

    let content: React.ReactNode;
    let isEditable = !!props.onValueChange;

    const handleEditorClick: React.MouseEventHandler<HTMLDivElement> = React.useCallback((e) => {
        const input: HTMLInputElement = (e.target as HTMLElement).querySelector('.' + uuiElement.input);
        input?.focus();
    }, []);

    if (props.rowProps.isLoading) {
        content = props.renderPlaceholder(props);
    } else if (isEditable) {

        // Copy all attributes explicitly, to avoid bypassing unnecessary DataTableCell props
        // We don't use any helpers and/or deconstruction syntax, as this is performance-sensitive part of code
        const editorProps: RenderEditorProps<TItem, TId, any> = {
            value: props.value,
            onValueChange: props.onValueChange,
            isFocused: state.inFocus,
            isDisabled: props.isDisabled,
            isInvalid: props.isInvalid,
            isReadonly: props.isReadonly,
            isRequired: props.isRequired,
            validationMessage: props.validationMessage,
            validationProps: props.validationProps,
            onFocus: () => setState({ ...state, inFocus: true }),
            onBlur: () => setState({ ...state, inFocus: false }),
            rowProps: props.rowProps,
            mode: 'cell',
        };

        content = <div
            className={ css.editorWrapper }
            onClick={ handleEditorClick }
        >
            { props.renderEditor(editorProps) }
            <DataTableCellOverlay
                { ...editorProps }
                renderTooltip={ props.renderTooltip }
                inFocus={ state.inFocus }
                rowIndex={ row.index }
                columnIndex={ props.index }
            />
        </div>;
    } else {
        content = props.column.render(props.rowProps.value, props.rowProps);
    }

    let justifyContent = props.column.justifyContent;
    if (!justifyContent && props.column.textAlign) {
        justifyContent = props.column.textAlign;
    }

    const { textAlign, alignSelf } = props.column;
    const styles = { textAlign, alignSelf, justifyContent };

    const getWrappedContent = () => (
        <div style={ styles } className={ css.contentWrapper }>
            { content }
        </div>
    );

    return (
        <FlexCell
            ref={ ref }
            grow={ props.column.grow }
            width={ props.column.width }
            minWidth={ props.column.width }
            textAlign={ props.isFirstColumn ? undefined : props.column.textAlign }
            alignSelf={ props.isFirstColumn ? undefined : props.column.alignSelf }
            rawProps={ { role: 'cell' } }
            cx={ [
                uuiDataTableCellMarkers.uuiTableCell,
                css.cell,
                props.column.cx,
                props.cx,
                props.isInvalid && uuiMod.invalid,
                state.inFocus && uuiMod.focus,
            ] }
            style={ !props.isFirstColumn && { justifyContent: justifyContent } }
        >
            { props.addons }
            { props.isFirstColumn ? getWrappedContent() : content }
        </FlexCell>
    );
};
