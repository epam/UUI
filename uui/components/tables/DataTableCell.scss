.cell {
    position: relative;
    box-sizing: border-box;
    width: 0;

    &.wrapper {
        display: flex;
        align-items: flex-start;
    }

    &.align-widgets-top {
        &.size-24, &.size-30 {
            :global(.uui-checkbox), .folding-arrow {
                margin-top: 6px;
            }
        }

        &.size-36 {
            :global(.uui-checkbox), .folding-arrow {
                margin-top: 9px;
            }
        }

        &.size-42 {
            :global(.uui-checkbox), .folding-arrow {
                margin-top: 12px;
            }
        }

        &.size-48 {
            :global(.uui-checkbox), .folding-arrow {
                margin-top: 15px;
            }
        }

        &.size-60 {
            :global(.uui-checkbox), .folding-arrow {
                margin-top: 21px;
            }
        }
    }

    &.align-widgets-center { align-self: center; }

    &.padding-12 { padding: 0 12px; }
    &.padding-24 { padding: 0 24px; }

    &.padding-left-12 { padding-left: 12px; }
    &.padding-left-24 { padding-left: 24px; }

    &.padding-right-24 { padding-right: 24px; }
}

.drag-handle {
    position: absolute;
    left: 2px;
    width: 6px;
    top: 3px;
    bottom: 2px;
    align-self: stretch;
    visibility: hidden;

    :global(.-draggable.uui-table-row):not(:global(.uui-drop-accepted)):hover & {
        visibility: visible;
    }
}

.checkbox {
    padding-right: 12px;
}

.indent {
    display: flex;
    justify-content: flex-start;
    flex-grow: 0;
    flex-shrink: 0;
    width: 12px;
    padding-right: 12px;
}

.folding-arrow {
    margin: auto 0;

    &:global(.uui-icon) {
        position: relative;
        fill: var(--uui-icon);
    }
}

.folding-arrow-12 {
    &:global(.uui-icon) {
        left: -2px;
        flex: 0 0 12px;
        width: 12px;
        height: 12px;
    }
}

.folding-arrow-18 {
    &:global(.uui-icon) {
        left: -4px;
        flex: 0 0 18px;
        width: 18px;
        height: 18px;
    }
}


