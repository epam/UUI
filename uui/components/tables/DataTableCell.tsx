import * as React from 'react';
import {
    DataTableCellProps as UuiCoreDataTableCellProps,
    Overwrite,
    uuiMod,
} from '@epam/uui-core';
import { DataTableCell as UuiDataTableCell } from '@epam/uui-components';
import { DataRowAddons, DataRowAddonsProps } from '../widgets';
import type { DataTableCellMods } from './types';
import { Text } from '../typography';
import { Tooltip } from '../overlays';
import { settings } from '../../settings';

import './variables.scss';
import css from './DataTableCell.module.scss';

export interface DataTableCellModsOverride {}

export interface DataTableCellProps<TItem, TId, TCellValue> extends
    UuiCoreDataTableCellProps<TItem, TId, TCellValue>, Pick<DataRowAddonsProps<TItem, TId>, 'renderDragHandle'>, Overwrite<DataTableCellMods, DataTableCellModsOverride> {
}

export function DataTableCell<TItem, TId, TCellValue>(initialProps : DataTableCellProps<TItem, TId, TCellValue>) {
    const props = { ...initialProps };

    if (props.isFirstColumn) {
        props.addons = <DataRowAddons size={ props.size } { ...props } />;
    }

    props.renderPlaceholder = props.renderPlaceholder
        || (() => settings.dataTable.renderPlaceholder({ rowSize: props.size }));

    props.renderUnknown = props.renderUnknown
        || (() => (
            <Text key="t" size={ props.size }>
                Unknown
            </Text>
        ));

    props.renderTooltip = (tooltipProps) => <Tooltip color="critical" { ...tooltipProps } />;

    const isEditable = !!props.onValueChange;

    const getLeftPadding = () => {
        const { rowProps: { isLoading }, columnsGap, isFirstColumn } = props;

        if (isFirstColumn && isEditable && !isLoading) return 'var(--uui-dt-cell-padding)';
        if (isEditable && !isLoading) return '0px';
        if (columnsGap) return isFirstColumn ? `${columnsGap}px` : `${+columnsGap / 2}px`;
        return `var(--uui-dt-cell-padding${isFirstColumn ? '-edge' : ''})`;
    };

    const getRightPadding = () => {
        const { rowProps: { isLoading }, columnsGap, isLastColumn } = props;

        if (isEditable && !isLoading) return '0px';
        if (columnsGap) return isLastColumn ? `${columnsGap}px` : `${+columnsGap / 2}px`;
        return `var(--uui-dt-cell-padding${isLastColumn ? '-edge' : ''})`;
    };

    props.cx = [
        'data-table-cell',
        css.root,
        props.cx,
        'uui-size-' + (props.size || settings.dataTable.sizes.body.row),
        props.isFirstColumn && 'uui-dt-first-column',
        props.isLastColumn && 'uui-dt-last-column',
        css[`align-widgets-${props.alignActions || 'top'}`],
        (props.border || isEditable) && 'uui-dt-vertical-cell-border',
        props.isDisabled && uuiMod.disabled,
        props.isReadonly && uuiMod.readonly,
    ];

    props.style = {
        '--uui-dt-cell-padding-start': getLeftPadding(),
        '--uui-dt-cell-padding-end': getRightPadding(),
    } as React.CSSProperties;

    return <UuiDataTableCell { ...props } />;
}
