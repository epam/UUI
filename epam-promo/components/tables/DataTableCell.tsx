import * as React from 'react';
import { cx, uuiMarkers } from '@epam/uui';
import { IconContainer, DragHandle } from '@epam/uui-components';
import { DataTableCellMods, DataTableCellProps } from './types';
import { TextPlaceholder, Text } from '../typography';
import { FlexCell } from '../layout';
import { Checkbox } from '../inputs';
import * as css from './DataTableCell.scss';
import * as foldingArrow from '../../icons/tree_folding_arrow.svg';

export type ItemsForRender = {
    component: React.ReactNode;
    widthCost: number;
};

export class DataTableCell extends React.Component<DataTableCellProps<any, any> & DataTableCellMods, {}> {
    getCellPadding = () => {
        if (this.props.padding) {
            return +this.props.padding;
        }
        return +this.props.size < 30 ? 18 : 24;
    }

    render() {
        const row = this.props.rowProps;
        let cellContent = row.isLoading
            ? <Text size={ this.props.size != '60' ? this.props.size : '48' }><TextPlaceholder /></Text>
            : this.props.column.render(this.props.rowProps.value, this.props.rowProps);


        let reusablePadding = this.props.reusePadding === 'false' ? 0 : this.getCellPadding(); // how much padding we can reuse for addon widgets. !Temporary left value from old design!
        const addonWidgets = []; // addon components, like checkbox, label, and folding arrow

        if (this.props.isFirstColumn) { // process addon widgets
            const additionalItemSize = +this.props.size < 30 ? '12' : '18';
            let allAddons: ItemsForRender[] = [];

            if (row.dnd) {
                allAddons.push({
                    component: <>{ row.dnd && row.dnd.srcData && <DragHandle cx={ css.dragHandle } /> }</>,
                    widthCost: 6,
                });
            }

            if (row.checkbox && row.checkbox.isVisible) {
                allAddons.push({
                    component: <Checkbox
                        key='cb'
                        cx={ css.checkbox }
                        size={ additionalItemSize }
                        value={ row.isChecked }
                        indeterminate={ !row.isChecked && row.isChildrenChecked }
                        onValueChange={ () => row.onCheck(row) }
                        isDisabled={ row.checkbox.isDisabled }
                        isInvalid={ row.checkbox.isInvalid }
                    />,
                    widthCost: 24,
                });
            }

            if (row.indent > 0) {
                allAddons.push({
                    component: <div key='fold' className={ css.indent } style={ { marginLeft: (row.indent - 1) * 24 } }>
                        { row.isFoldable && <IconContainer
                            key='icon'
                            icon={ foldingArrow }
                            cx={ [css.foldingArrow, css[`folding-arrow-${additionalItemSize}`], uuiMarkers.clickable] }
                            rotate={ row.isFolded ? '90ccw' : '0' }
                            onClick={ () => row.onFold(row) }
                        /> }
                    </div>,
                    widthCost: row.indent * 24,
                });
            }

            // Split all addons in two parts - which are placed in padding, and which are placed in the cell
            let addonsInPadding: ItemsForRender[] = [];
            let addonsInCell: ItemsForRender[] = [];

            allAddons.forEach(item => {
                if (item.widthCost < reusablePadding) {
                    addonsInPadding.push(item);
                    reusablePadding -= item.widthCost;
                } else {
                    addonsInCell.push(item);
                }
            });

            if (addonsInPadding.length > 0) {
                addonWidgets.push(<div key='bp' className={ css.beforePaddings } style={ { 'width': `12px` } }>
                    { addonsInPadding.map(item => item.component) }
                </div>);
            }

            if (addonsInCell.length > 0) {
                addonWidgets.push(<div key='ap' className={ css.afterPaddings }>{ addonsInCell.map(a => a.component) }</div>);
            }
        }

        // If we add any addons, we need to place them, as well as user-defined content in a flexbox.
        // To keep user's layout in clock context (not in flex), we wrap it in a div that spans over whole available space.
        if (addonWidgets.length > 0) {
            cellContent = <div key='w' className={ css.wrappedContent }>{ cellContent }</div>;
        }

        return (
            <FlexCell
                { ...this.props.column }
                cx={ cx(
                    css.cell,
                    addonWidgets.length > 0 && css.wrapper,
                    css['size-' + (this.props.size || '36')],
                    css[`padding-${ this.props.padding || '12' }`],
                    this.props.isFirstColumn && css[`padding-left-${ this.getCellPadding() }`],
                    this.props.isLastColumn && css['padding-right-24'],
                    this.props.column.cx,
                    css[`align-widgets-${ this.props.alignActions || 'top' }`],
                ) }>
                { addonWidgets }
                { cellContent }
            </FlexCell>
        );
    }
}