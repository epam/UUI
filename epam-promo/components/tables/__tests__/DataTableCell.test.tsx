import React from 'react';
import { DataTableCell } from '../DataTableCell';
import renderer from 'react-test-renderer';

describe('DataTableCell', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <DataTableCell
                    column={{
                        key: 'test',
                        caption: 'Test',
                        render: () => <div>Test</div>,
                        width: 150,
                        fix: 'left',
                    }}
                    rowProps={{
                        id: '1',
                        rowKey: '1',
                        index: 1,
                        dnd: {
                            canAcceptDrop: params => null,
                            onDrop: jest.fn,
                            srcData: {},
                            dstData: {},
                        },
                    }}
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <DataTableCell
                    column={{
                        key: 'test',
                        caption: 'Test',
                        render: () => <div>Test</div>,
                        width: 150,
                        fix: 'left',
                        info: 'test',
                        isSortable: true,
                    }}
                    rowProps={{
                        id: '1',
                        rowKey: '1',
                        index: 1,
                        dnd: {
                            canAcceptDrop: params => null,
                            onDrop: jest.fn,
                            srcData: {},
                            dstData: {},
                        },
                    }}
                    isLastColumn
                    padding="24"
                    size="48"
                    reusePadding="auto"
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
