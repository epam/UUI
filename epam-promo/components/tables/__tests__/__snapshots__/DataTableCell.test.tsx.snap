// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableCell should be rendered correctly 1`] = `
<div
  className="cell size-36 padding-12 align-widgets-top"
  style={
    Object {
      "alignSelf": undefined,
      "flexBasis": "150px",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "minWidth": "150px",
      "textAlign": undefined,
    }
  }
>
  <div>
    Test
  </div>
</div>
`;

exports[`DataTableCell should be rendered correctly 2`] = `
<div
  className="cell size-48 padding-24 padding-right-24 align-widgets-top"
  style={
    Object {
      "alignSelf": undefined,
      "flexBasis": "150px",
      "flexGrow": undefined,
      "flexShrink": undefined,
      "minWidth": "150px",
      "textAlign": undefined,
    }
  }
>
  <div>
    Test
  </div>
</div>
`;
